### PR TITLE
Once and for all deal with dockerpy's brokenness

### PR DESCRIPTION
**Pull request by chuckbutler**
This adds a custom build tactic largely recommended by coryfu, however, it needs additional work to be a clean solution.

This requires some refactoring in charm-tools to clean up the wheelhouse tactic, and split into a few different methods so this becomes a 3 line tactic.

For now though, this unblocks all work with layer-docker and allows us to continue including upstream dependencies

### CODE REVIEW

File: tactics/docker.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class DockerWheelhouseTactic(WheelhouseTactic):
+    def __call__(self, venv=None):
+        from charmtools import utils
+        create_venv = venv is None
+        venv = venv or path(tempfile.mkdtemp())
+        pip = venv / 'bin' / 'pip3'
+        wheelhouse = self.target.directory / 'wheelhouse'
+        wheelhouse.mkdir_p()
+        if create_venv:
+            import pdb; pdb.set_trace()
```
Comment (mbruzek) [inline tactics/docker.py]: remove the pdb

File: tactics/docker.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from charmtools.build.tactics import WheelhouseTactic
+
+
+class DockerWheelhouseTactic(WheelhouseTactic):
+    def __call__(self, venv=None):
+        from charmtools import utils
```
Comment (mbruzek) [inline tactics/docker.py]: Why do you import utils here and not at top of file?

File: tactics/docker.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        create_venv = venv is None
+        venv = venv or path(tempfile.mkdtemp())
```
Comment (mbruzek) [inline tactics/docker.py]: I don't see you import path.py anywhere. Explain to me how this works.

File: tactics/docker.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+from charmtools.build.tactics import WheelhouseTactic
+
+
+class DockerWheelhouseTactic(WheelhouseTactic):
+    def __call__(self, venv=None):
+        # this was due to some weird scoping issue during execution.
+        # investigate more later
+        from charmtools import utils
+        create_venv = venv is None
+        venv = venv or path(tempfile.mkdtemp())
```
Comment (mbruzek) [inline tactics/docker.py]: I don't see you importing path.py here. How does this work?

Comment (chuckbutler) [inline tactics/docker.py]: I'm not entirely sure, as I had to import utils 2 lines above that.

I do believe however that venv is being set before hand and being passed in. As temple is not included either and would throw an error.

**Comment by mbruzek** [comment]
Using modules you didn't explicitly import. This looks like witchcraft to me!

I would like @bcsaller or @johnsca to comment on this code and why it works.

File: tactics/docker.py (hunk 4 above)
Comment (marcoceppi) [inline tactics/docker.py]: basically, if tactics don't pass in venv this will fail because of lack of path.

**Comment by mbruzek** [comment]
+1 LGTM. Thanks for making the code more complete by importing the libraries you are using.

File: tactics/docker.py (hunk 4 above)
Comment (mbruzek) [inline tactics/docker.py]: Actually @marcoceppi we found out that the utils.load_class() method was loading this library with it's global namespace/libraries and we were able to use path, tempfile and other libraries without importing them.

**Pull request merged**